First working prototype - DHCPv4

yoelcaspersen opened the pull request:
This version works with kernel 5.15 and LLVM 13.

Improvements:

- Interface name is added to Option 82 Circuit ID
- On server replies, Option 82 is wiped completely
- Char array parameters replaced with structs

Functions that previously used char arrays as function parameters now
use structs instead - otherwise the verifier cannot perform proper
boundary checks.

To-do:
- IPv6 support
- Multiple interface support

Commit:
Adding custom memcpy and memset functions to pass verifier checks
Built-in memcpy and memset functions cause verifier to complain about
misaligned stack access. To get around that (and be able to copy a
dynamic number of bytes via memcpy), two custom functions are added.
This version of the relay uses interface name + VLAN identifiers in
option 82 circuit ID (left-aligned, length IF_NAMESIZE with trailing
null bytes).

Signed-off-by: Yoel Caspersen <yoel@kviknet.dk>

Commit:
Upstream direction works with QinQ packets only
When allowing packets with 1 VLAN tag only, verifier complains about
BPF program being too large.
Compiled with LLVM 13.

Signed-off-by: Yoel Caspersen <yoel@kviknet.dk>

tohojo commented on dhcp-relay/Makefile:
```diff
@@ -4,6 +4,7 @@ USER_TARGETS := dhcp_user_xdp
 BPF_TARGETS :=dhcp_kern_xdp 
 EXTRA_DEPS := dhcp-relay.h
 #EXTRA_CFLAGS := $(if $(IPV6),-DIPV6)
+EXTRA_CFLAGS := -fno-builtin
```
It would be better if we could build without this...

yoelcaspersen replied:
Sure, but then we'll need to find a way to copy an arbitrary number of bytes without having the compiler default to memcpy().

tohojo commented on dhcp-relay/dhcp-relay.h:
```diff
@@ -10,19 +10,21 @@
 #define RAI_CIRCUIT_ID 1
 #define RAI_REMOTE_ID 2
-#define RAI_OPTION_LEN 40
+#define RAI_OPTION_LEN IF_NAMESIZE
```
Is this really long enough? Theoretically the interface itself can be IF_NAMESIZE long?

yoelcaspersen replied:
Right now I am just trying to make it work with an interface name that contains the VLAN tags as part of the name (and therefore is at most IF_NAMESIZE long). Once it works with 16 bytes, increasing to e.g. 32 or more should simply be a matter of increasing RAI_OPTION_LEN.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
 	__uint(max_entries, 16384);
 } client_vlans SEC(".maps");
 
+void memcpy_var(void *to, void *from, __u64 len) {
```
With MAX_LOOPS this can fail, so it should probably no longer be void, and instead return an error code if it does...

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
+	__u8 *t8 = to, *f8 = from;
+	int i;
+
+	for (i = 0; i < len && i < MAX_LOOPS; i++) {
```
I think it should be enough to just add a check for len > MAX_LOOPS (and return error if it is)?

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
+		
+}
+
+void memset_var(void *d, __u8 c, __u64 len) {
```
Same as above

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
 static __always_inline int write_dhcp_option_82(void *ctx, int offset,
 	struct collect_vlans *vlans, char *dev) {
+	
 	struct dhcp_option_82 option;
```
Why not just zero out the whole struct with an initialiser (`= {}` here) instead of the not-quite-_var memsets below?

tohojo replied:
Or even initialise the options directly using {} syntax...

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
 	struct dhcp_option_82 option;
-
+	
+	static __u8 buf[RAI_OPTION_LEN];
```
This can also be initialised to 0 instead of the memset below

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
+	memset(buf, 0, sizeof (buf));
+	
+	int c = VLAN_ASCII_MAX; /* We will need 4 bytes at most */
+	int i = RAI_OPTION_LEN - 1;
```
sizeof(buf)?

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
-	
+
+	/* Convert inner VLAN to ASCII */
+#pragma unroll VLAN_ASCII_MAX
```
Is this really needed?

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
 		if (inner_vlan == 0) {
 			break;
 		}
 	}
```
As I said elsewhere, this should be a helper function

yoelcaspersen replied:
Will it decrease the resulting eBPF object size or make it easier to pass the verifier?

Bear in mind, convert an int between 0 and 4095 to ASCII is a very specific thing to do...

tohojo replied:
...but you're doing it twice? That alone merits a helper :)

As for decreasing complexity, yes if the functions are *not* inlined, the verifier should be able to do per-function verification, which can decrease overall complexity. Probably not worth it for this helper, but the full write_option82 helper could be a contender for this.

yoelcaspersen replied:
Helper function u16_to_ascii(char *buf, __u8 offset, __u16 num) added now.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
+	/* Append interface name */
+#pragma unroll RAI_OPTION_LEN
+	for (c = RAI_OPTION_LEN - 1; c >= 0; c--) {
+		if (dev[c] != 0)
```
`if (dev[c] == 0) break;` ?

yoelcaspersen replied:
It is counting down - the first couple of bytes are more than likely to be null. It needs to skip the null bytes before copying the actual interface name.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
+	i++;
+	
+	/* Copy resulting interface name to circuit_id */
+	if (sizeof (option.circuit_id.val) == sizeof (buf)) {
```
Why this check?

yoelcaspersen replied:
The only way I could make it pass the verifier.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,76 +56,104 @@ struct {
+		//memcpy_var(option.circuit_id.val, buf, sizeof (buf));
+	}
 	
 	return xdp_store_bytes(ctx, offset, &option, sizeof (option), 0);
```
The option should be resized to the right size based on the string length as well, no?

yoelcaspersen replied:
Yes, ideally it should, but as far as I can see, that would rule out the use of xdp_store_bytes(). Copying dynamic lengths of memory really is an uphill battle with XDP.

The current version actually works in the sense that it can send a packet to the DHCP server and receive a reply. But the verifier explodes if I disable the check that skips single tagged packets (like the reply that comes from the server):

```
	if (vlans.id[1] == 0) {
		bpf_printk("No inner VLAN tag set\n");
		goto out;
	}
```
The verifier complains that the BPF program is too large. I guess it cannot cope with the loops.

tohojo replied:
My guess would be that this check allows the compiler to make some assumptions that changes the code flow, and removing it makes things longer. Or something along those lines. The above-mentioned per-function verification might be a way around this...

yoelcaspersen replied:
Apparently the primary issue was use of char arrays as functions parameters. After an upgrade to kernel 5.15 I replaced the char arrays with pointers to structs containing the necessary char arrays. 

That seems to work, as long as just the right amount of boundary checks are littered across the code.

Commit:
First working prototype (IPv4 DHCP relay works in both directions)

In this version of the relay, we assume that client requests are
received on a double tagged VLAN interface and server replies are
received on a single tagged VLAN interface.

Test scripts are added:

- test.sh (configures BNG and sets L3 config)
- cleanup_test.sh (removes BNG configuration)
- trace.sh (opens trace pipe for debugging)

The code for adding interface name is disabled because it makes the
verifier complain about the BPF program being too large.

To-do:
- Fix verifier issue to include interface name in option 82
- For server replies, erase option 82 instead of (partial) overwrite

Signed-off-by: Yoel Caspersen <yoel@kviknet.dk>

Commit:
Working prototype (IPv4) with interface name in Option 82 Circuit ID

This version works with kernel 5.15 and LLVM 13.

Improvements:

- Interface name is added to Option 82 Circuit ID
- On server replies, Option 82 is wiped completely
- Char array parameters replaced with structs

Functions that previously used char arrays as function parameters now
use structs instead - otherwise the verifier cannot perform proper
boundary checks.

To-do:
- IPv6 support
- Multiple interface support

Signed-off-by: Yoel Caspersen <yoel@kviknet.dk>

tohojo commented on dhcp-relay/dhcp-relay.h:
```diff
@@ -38,6 +43,11 @@ struct dhcp_option_255 {
 
+struct dev_name {
+	char name[IF_NAMESIZE];
+	__u8 len;
```
len is unneeded, see below

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
+//
+//	return 0;
+//
+//}
```
Please remove commented-out code

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
+//
+//}
+
+int u16_to_ascii(struct sub_option *opt, __u8 offset, __u16 num) {
```
This is q fairly small helper function to make global; better keep this 'static' so it can be inlined, and make the whole `write_dhcp_option_82` function below global. Just having these helpers global are not enough to pass the verifier on my kernel either.

Also, since there are two different functions with an 'offset' parameter, but with different naming, rename the 'offset' parameter to 'end_offset'.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
+	}
+
+	if (i > 0) {
+		opt->val[--i] = '.';
```
The function is called `u16_to_ascii`, so what is this doing here? Better leave this to the caller.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
+	}
+
+	if (i > RAI_OPTION_LEN)
+		return -1;
```
This check is not needed...

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
+
+		opt->val[i] = (__u8) (num % 10) + '0';
+		num /= 10;
+		if (num == 0)
```
Right now the value of `i` is different depending on which condition exited the loop: if this one matches, `i` will be the last character written, but if the loop condition itself matches, `i` will be right before the last character written (i.e., -1). The code below assumes this is always the condition that matches (which it will be, in practice), but that is not clear from reading the code; so better to move this check into the loop condition as well, so `i` always ends up right before the last character written. This also requires setting `i = offset` above instead of `i = offset - 1`, as then things fit when passing `i` back in for multiple calls.

tohojo replied:
Actually, this difference is fine, as writing to the first character of the string without having `num` be 0 afterwards *should* be an error (i.e., -1 return); and moving the '.' assignment out of the function removes the assumption that this was the check that exited the loop. However, there's an off-by-one error in the loop condition above, it should be `i >= 0` (as writing to the first byte of the buffer is fine).

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
+
+}
+
+int str_len(struct dev_name *dev) {
```
This is not needed either, see below

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
+			/* Move VLAN part (all bytes from offset and up) */
+			dest->val[i] = dest->val[offset];
+			dest->val[offset++] = 0;
+		}
```
This weird double-construction supposes a particular construction of the buffer being written to, which is hard to understand and makes the code less readable (and less reusable as well). Instead, how about turning this function into one that actually does what the name applies, like:

```c
static int copy_dev_name(struct sub_option *dest, __u8 wr_offset,
			 struct dev_name *dev, __u8 rd_offset) {
	int i;

	if (dest == NULL || dev == NULL || wr_offset > sizeof(dest->val) -1)
		return -1;

	for (i = 0; i < sizeof(dest->val) - wr_offset; i++) {
		if (i + rd_offset > sizeof(dev->name) ||
		    !dev->name[i + rd_offset])
			break;

		dest->val[i + wr_offset] = dev->name[i + rd_offset];
	}

 	return i;
}
```

and calling it like:

```c
	len = copy_dev_name(&option.circuit_id, 0, dev, 0);
	if (len < 0)
		return len;

	len = copy_dev_name(&option.circuit_id, len, &buf, vlan_offset);
	if (len < 0)
		return len;
```

where `buf` is a separate `struct dev_name` allocated on the stack, into which the VLAN tags are stringified.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
 /* Inserts DHCP option 82 into the received DHCP packet
  * at the specified offset.
  */
 static __always_inline int write_dhcp_option_82(void *ctx, int offset,
```
As mentioned above, better to make this one the global function; that requires turning 'dev' into a pointer argument (pass-by-value is not supported). I'd also suggest renaming `offset` to `pkt_offset`, as we already have way too many `offset`s in this code...

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
-	int c = VLAN_ASCII_MAX;		/* We will need 4 bytes at most */
-	int i = IF_NAMESIZE - 1;
+
+	int i = RAI_OPTION_LEN;
```
The kernel style guidelines forbid mixing variable declarations and code, so please move the declarations to the top of the function. Also, `i` is a terrible variable name (only use that for idiomatic loops), maybe `vlan_offset`? Finally, you can use `sizeof(buf.name)` (and changing the `u16_to_ascii` calls to this `struct dev_name buf` as mentioned above.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
-		if (outer_vlan == 0) {
-			break;
+
+	if (inner_vlan != 0) {
```
Everything seems to assume there are actually two VLAN tags present, so probably better to error out if either of the tags are unset, instead of this conditional (same below, of course).

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -49,77 +56,162 @@ struct {
-	option.remote_id.len = sizeof(option.remote_id.val);
-	
+	/* Insert device name and left-align circuit ID */
+	i = copy_dev_name(&option.circuit_id, i, &dev);
```
See above for this

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -136,10 +228,10 @@ static __always_inline int write_dhcp_option_255(void *ctx, int offset) {
 /* Calculates the IP checksum */
 static __always_inline int calc_ip_csum(struct iphdr *oldip, struct iphdr *ip,
-	__u32 oldcsum) {
+		__u32 oldcsum) {
```
Some odd alignment stuff going on here; generally, for multi-line function definitions, align the continuation of the parameters with the opening parenthesis on subsequent lines.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -166,27 +258,18 @@ static __always_inline int calc_ip_csum(struct iphdr *oldip, struct iphdr *ip,
 	if (res != 0) {
 		bpf_printk("Cannot tail extend packet, delta %i - error code %i", delta, res);
-		return XDP_ABORTED;
+		return XDP_PASS;
```
This is a packet we figured we should handle, but that failed; so what is XDP_PASS supposed to do with it? Better to drop...

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -217,34 +302,38 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
-		return XDP_ABORTED;
+	if (data + 1 > data_end) {
+		goto out;
+	}
```
Unneded braces

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -217,34 +302,38 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
+
 	if (ether_type != bpf_htons(ETH_P_IP)) {
-		bpf_printk("Ethertype %#x is not ETH_P_IP", bpf_ntohs(ether_type));
+		//bpf_printk("Ethertype %x is not ETH_P_IP", bpf_ntohs(ether_type));
```
Remove commented-out code...

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -256,15 +345,19 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 	/* Increase UDP length header */
 	udp->len += bpf_htons(delta);
 
+	udp->check = 0;
```
Here you're modifying the packet, but any `goto out` statements below will end up returning `XDP_PASS`, passing up a potentially partially-modified packet. `rc` should be changed to `XDP_DROP` at the point of the first packet modification.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -285,23 +378,20 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
-	//if (xdp_load_bytes(ctx, 0, buf, static_offset))
-	//	goto out;
+	memcpy(dev.name, dev_config, RAI_OPTION_LEN);
+	dev.len = str_len(&dev);
```
This is just looping through the string to find the length of it, only to use that length as the end condition of a loop later. This is totally superfluous, just check for the null terminator directly in the loop that uses length and get rid of this call (and lose the `len` member of struct dev_name entirely).

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -285,23 +378,20 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
+	dev.len = str_len(&dev);
 
 	/* Increment offset by 4 bytes for each VLAN (to accomodate VLAN headers */
+#pragma unroll VLAN_MAX_DEPTH
```
Not needed

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -317,6 +407,16 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
+		bpf_printk("Max hops exceeded, discarding packet");
+		rc = XDP_ABORTED;
+		goto out;
+	}
```
Now we're suddenly doing parsing and sanity checks of the packet in the middle of modifying it. Better to parse and check everything first, *then* start modifying.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -326,11 +426,11 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
+			&& eth->h_dest[2] == 0xff
+			&& eth->h_dest[3] == 0xff
+			&& eth->h_dest[4] == 0xff
+			&& eth->h_dest[5] == 0xff)) {
```
This calls for an `is_ether_broadcast()` helper (which can then be inverted below)

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
+		ip->daddr = IP_ADDR_BCAST;
+
+		/* Set source IP */
+		ip->saddr = *relay_agent_ip;
```
Jumping back and forth between modifying L2 and L3 headers makes the code hard to follow. Also this function is way too long, it should be split: let's turn it into separate `handle_request()` and `handle_reply()` functions, and also split the VLAN tag insertion/removal into a separate helper.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 
 	option_offset = offset;
 
+	__u8 n = 0;
```
See above cf mixing declarations and code. Also, like `i`, `n` is a terrible variable name outside very narrowly-scoped counter contexts.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 
 	/* Loop through all DHCP options */
-#pragma unroll DHCP_MAX_OPTIONS
+	//#pragma unroll DHCP_MAX_OPTIONS
```
Remove

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 		if (pos + 1 > data_end)
 			break;
 
+		/* Read option code */
```
Redundant comment; it just says what the code is doing...

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 
 		bpf_printk("Got option code %i at offset %i, hex %x", option_code, option_offset, option_offset);
 
+		if (option_code == 82 && dhcp->op == DHCP_REPLY) {
```
This code is wrong for several reasons:

- It doesn't handle option 82 for REQUESTS (it should drop the packet)
- It just blindly turns option 82 into an option 255, regardless of where it is in the packet (what if it's not actually the last option)
- It assumes that the option 82 is always the same size as the one we put in
- It leaves junk in the packet

The right thing to do is to shrink the packet; and either move everything up, including any non-255 options after this, or at least checking to see if there is anything other than an end option in the rest of the packet (and dropping it as invalid if there is).

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 		option_length = *pos;
 		option_offset += option_length + 2;
 
+		/* Verifier check */
```
These repeated 'verifier check' comments are just noise. There are also way too many checks one after another at the end here, some of them could definitely be collapsed. This may be easier to figure out after some of the other refactoring mentioned above.

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 		}
 		pos++;
 
+		/* Verifier check */
```
For instance, the check below could be moved up, and changed to `if (pos + option_length + 1 > data_end)`

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
-}
+	/* Adjust IP offset for VLAN header when VLAN header has been added */
+	if (head_adjusted) {
+		ip = data + ip_offset + sizeof (struct vlan_hdr);
```
Why carry this `head_adjusted` all the way down here instead of just modifying `ip_offset` directly when doing the adjustment?

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
-	if (ip + 1 > data_end)
+	if (ip + 1 > data_end) {
 		return XDP_ABORTED;
+	}
```
Unnecessary braces

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -499,7 +685,6 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 	ip->check = ~sum;
-	rc = XDP_PASS;
 
 	goto out;
```
Redundant goto

tohojo commented on dhcp-relay/dhcp_kern_xdp.c:
```diff
@@ -499,7 +685,6 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 	/* Re-calculate ip checksum */
 	__u32 sum = calc_ip_csum(&oldip, ip, oldip.check);
 	ip->check = ~sum;
-	rc = XDP_PASS;
```
See above; this *is* actually the right place to change the return code back into something that sends out the packet, as this is where the packet is once again valid...

tohojo commented on dhcp-relay/test.sh:
```diff
@@ -0,0 +1,65 @@
+OUTER_VLAN=83
+INNER_VLAN=20
+UPLINK_VLAN=84
+IF="ens6f0"
```
It should be possible to test this using network namespaces; having a test that is runnable without needing special hardware makes it way easier to test. Such a test should really be added so it can be run as part of a CI job.

tohojo reviewed (changes requested):
Nice job getting something to work! Unfortunately the code has turned a wee bit spaghetti in the process; I guess most of the comments below could be summed up as changes that fix that :)

tohojo committed:
Improve readability of write_dhcp_option_82()

This makes several changes to improve the readability of write_dhcp_option_82():

The main change is in reworking the string helpers to be smaller and more
modular, and using a separate buffer for stringifying the VLAN tags before
copying them to the DHCP option field. Get rid of str_len() and the len
member of struct dev_name. To aid in this, the whole containing function is
made global (i.e., a separate BPF subprog) instead of the helpers.

In addition, the local variables are renamed to be more descriptive, and
the option struct is initialised using struct initialisation syntax.

Signed-off-by: Toke Høiland-Jørgensen <toke@redhat.com>

tohojo committed:
dhcp_user_xdp: Fix indentation of switch cases

The case statements of the option parsing 'switch' had grown an extra level
of indentation. Get rid of that.

Signed-off-by: Toke Høiland-Jørgensen <toke@redhat.com>

tohojo committed:
dhcp_user_xdp: Add -v option to enable verbose logging

Add a -v option to turn on verbose logging in libbpf; this makes it
possible to see which relocations libbpf is making, which is useful to see
which functions get turned into separate sub functions.

Signed-off-by: Toke Høiland-Jørgensen <toke@redhat.com>

tohojo commented:
Took the liberty of fixing some of the issues I pointed out myself, and pushed them to your branch :)
